pkg/pulumiyaml/run.go

```diff
@@ -176,16 +176,138 @@ func HasDiagnostics(err error) (syntax.Diagnostics, bool) {
 	}
 }
 
+func setDefaultProviders(ctx *pulumi.Context, t *ast.TemplateDecl, r *runner) error {
+	defaultProviderInfoMap := make(map[string]*providerInfo)
+
+	for _, resource := range t.Resources.Entries {
+		v := resource.Value
+		// check if this is a provider resource
+		if strings.HasPrefix(v.Type.Value, "pulumi:providers:") {
+			pkgName := strings.Split(v.Type.Value, "pulumi:providers:")[1]
+			// check if it's set as a default provider
+			if v.DefaultProvider != nil && v.DefaultProvider.Value {
+				defaultProviderInfoMap[pkgName] = &providerInfo{
+					version:           v.Options.Version,
+					pluginDownloadUrl: v.Options.PluginDownloadURL,
+					providerName:      resource.Key,
+				}
+			}
+		} else if v.DefaultProvider != nil {
+			return errors.New("cannot set defaultProvider on non-provider resource")
+		}
+	}
+
+	// Set roots
+	diags := r.Run(walker{
+		VisitResource: func(r *runner, node resourceNode) bool {
+			v := node.Value
+			ctx := r.newContext(node)
+			if strings.HasPrefix(v.Type.Value, "pulumi:providers:") {
+				return true
+			}
+			pkgName := strings.Split(v.Type.Value, ":")[0]
+
+			if _, ok := defaultProviderInfoMap[pkgName]; !ok {
+				return true
+			}
+			defaultProviderInfo := defaultProviderInfoMap[pkgName]
+
+			if v.Options.Provider == nil {
+				if v.Options.Version != nil {
+					if defaultProviderInfo.version.Value != v.Options.Version.Value {
+						ctx.ctx.Log.Error("version set conflicts with default provider version", &pulumi.LogArgs{})
```

Comment:
I think instead of `ctx.ctx.Log.Error`, you should be able to call `ctx.errorf` in these situations, where the first argument is the expression where you detected an error.

I'd also simplify the logic here: if there is a default provider _period_, the user shouldn't be setting the version on resources. 

```go
    ctx.errorf(v.Options.Version, "Version conflicts with the default provider version. Try removing this option on resource %q.", resource name, ... etc)
```

Reply:
& a nit: it'd be good to be helpful and prescribe remedies in error messages, such as deleting the option.